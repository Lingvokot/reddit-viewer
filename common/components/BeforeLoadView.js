import React, { Component } from "react";
import { Button, View, Text } from "react-native";

export default class BeforeLoadView extends Component {
	render() {
		if (!this.props.finished)
			return (
				<Text>Loading posts, please wait...</Text>
			);
		else if (this.props.error)
			return (
				<View>
					<Text>{this.props.error}</Text>
					<Button title="Retry to load"
							onPress={this.props.retryFunc}
							style={{
								paddingTop: 5, paddingBottom: 5,
								margin: 5
							}} ref={(ref) => {this.retryButton = ref;}}/>
				</View>
			);
		else
			return null;
	}
}