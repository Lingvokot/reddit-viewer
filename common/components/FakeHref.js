import React, { Component, PropTypes } from "react";
import { TouchableHighlight, Linking, Text } from "react-native";

export default class FakeHref extends Component {
	render() {
		return (
			<TouchableHighlight
				onPress={() => {
					return Linking.canOpenURL(this.props.url).then(supported => {
						if (!supported)
							this.props.onCannotHandle();
						else
							return Linking.openURL(this.props.url);
					}).catch(this.props.errorHandler);
				}} ref={(ref) => {this.highlight = ref;}}>
				<Text style={{color: "blue", textDecorationLine: "underline"}}>
					{this.props.url}
				</Text>
			</TouchableHighlight>
		);
	}
}
