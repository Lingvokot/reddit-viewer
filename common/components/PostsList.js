import React, { Component, PropTypes } from 'react';
import { View, Text, TouchableHighlight, ListView } from 'react-native';

export default class PostsList extends Component {
	render() {
		this.listElements = [];
		if (this.props.toList.getRowCount() == 0)
			return (
				<Text>There are no posts yet.</Text>
			)
		return (
			<ListView dataSource={this.props.toList}
				onEndReached={() => {
					this.props.onEndReached();
				}}
				renderRow={(rowData) => (
					<TouchableHighlight style={{
							borderWidth: 1,
							borderRadius: 3,
							borderColor: "red",
							margin: 5,
							padding: 5
						}} ref={(ref) => {this.listElements.push(ref);}}
						onPress={() => {this.props.onForward(rowData);}}>
						<Text>{rowData.title}</Text>
					</TouchableHighlight>
				)} ref={(ref) => {this.listView = ref;}}/>
		);
	}
}