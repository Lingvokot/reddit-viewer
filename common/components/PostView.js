/*global alert*/
import React, { Component, PropTypes } from "react";
import { View, Text, Button, Modal } from "react-native";
import FakeHref from "./FakeHref";

export default class PostView extends Component {
	constructor(props) {
		super(props);
		this.state = {urlOpenError: null};
	}
	render() {
		let selectedPost = this.props.selectedPost;
		return (
			<View>
				<Button onPress={this.props.onBack}
						title="Back"
						style={{
							paddingTop: 5, paddingBottom: 5,
							backgroundColor: "blue"
						}} ref={(ref) => {this.backButton = ref;}}/>
				<Text style={{fontWeight: "bold", textAlign: "center"}}>
					{selectedPost.title}
				</Text>
				<Text>{`ID: ${selectedPost.name}`}</Text>
				<Text>{`Author: ${selectedPost.author}`}</Text>
				<Text>{`Score: ${selectedPost.score}`}</Text>
				<Text>External URL:</Text>
				<FakeHref url={selectedPost.cleanUrl}
					onCannotHandle={() => {
						alert("Cannot handle this URL");
					}}
					errorHandler={(err) => {
						alert(err.message);
					}} ref={(ref) => {this.externalLink = ref;}}/>
			</View>
		);
	}
}