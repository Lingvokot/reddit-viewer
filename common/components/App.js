import React, { Component } from 'react';
import { Navigator, ListView } from 'react-native';
import PostsList from './PostsList';
import PostView from './PostView';
import BeforeLoadView from './BeforeLoadView';
import { last } from 'lodash/array';
import { each } from 'lodash/collection';
import loadPosts from './loadPosts';

const cardsInPage = 10;

export default class App extends Component {
	constructor(props) {
		super(props);
		this.state = {
			selectedPost: null, postsToList: [], pagesFetched: 0,
			error: null, finished: false, inProcess: false
		};
	}
	componentWillMount() {
		if (this.state.postsToList.length == 0)
			this.loadMorePosts();
	}
	loadMorePosts() {
		if (this.state.inProcess)
			return;
		return new Promise((resolve) => {
			this.setState({finished: false, inProcess: true}, async () => {
				let page = null;
				try {
					if (this.state.postsToList.length == 0)
						page = await loadPosts({limit: 10});
					else {
						const after = last(this.state.postsToList).uuid;
						page = await loadPosts({after, limit: 10});
					}
				}
				catch (e) {
					this.setState({error: e.message, finished: true, inProcess: false}, () => resolve());
					return;
				}
				let newPosts = Object.values(page.apiResponse.posts);
				this.setState({
					error: null, finished: true, inProcess: false,
					postsToList: this.state.postsToList.concat(newPosts),
					pagesFetched: this.state.pagesFetched + 1
				}, () => resolve());
			});
		});
	}
	render() {
		if (this.state.error || (!this.state.finished && this.state.postsToList.length == 0))
			return (
				<BeforeLoadView error={this.state.error}
					finished={this.state.finished}
					retryFunc={() => this.loadMorePosts()}/>
			)
		return (
			<Navigator initialRoute={{ title: 'Posts', index: 0 }}
				renderScene={(route, navigator) => {
				const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
				if (route.index == 0)
					return (
						<PostsList toList={ds.cloneWithRows(this.state.postsToList)}
							onForward={(selectedPost) => {
								return new Promise(resolve => {
									this.setState({selectedPost});
									resolve();
									navigator.push({title: selectedPost.title, index: route.index + 1});
								}).catch(err => {});
							}}
							onEndReached={() => this.loadMorePosts()}
							ref={(ref) => {this.postsList = ref;}}/>
					)
				else
					return (
						<PostView selectedPost={this.state.selectedPost}
							onBack={() => {
								return new Promise(resolve => {
									this.setState({selectedPost: null});
									resolve();
									navigator.pop();
								}).catch(err => {});
							}} ref={(ref) => {this.postView = ref;}}/>
					)
			}} />
		);
	}
}