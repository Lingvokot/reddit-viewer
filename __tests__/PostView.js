/*global jest, describe, it, test, jasmine, expect, console*/
/*eslint no-console: "off"*/
import "react-native";
import {ListView, Linking, Alert} from "react-native";
import React from "react";
import Index from "../index.android.js";
import App from "../common/components/App";
import BeforeLoadView from "../common/components/BeforeLoadView";
import PostView from "../common/components/PostView";
import PostsList from "../common/components/PostsList";
import FakeHref from "../common/components/FakeHref";
const {configureMock, makeArtificialErrorEvent} = Linking;

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
jest.useFakeTimers();
jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;

describe("Testing PostView", () => {
	const makeTestFunction = (expectedAlertStr, canOpenURL, errorHandling) => {
		return (done) => {
			configureMock(canOpenURL, errorHandling);
			console.warn({expectedAlertStr, canOpenURL, errorHandling});
			const highlight = domTree.getInstance().externalLink.highlight;
			highlight.props.onPress().then(() => {
				console.warn("butthurt1");
				if (expectedAlertStr) {
					expect(Alert.alert.mock.calls.length).toEqual(1);
					expect(Alert.alert.mock.calls[0]).toEqual([expectedAlertStr]);
				}
				else {
					expect(Alert.alert.mock.calls.length).toEqual(0);
				}
				done();
			});
		};
	};
	const selectedPost = {
		"cleanUrl":"http://www.businessinsider.com.au/elijah-wood-hollywood-pedophile-ring-2016-5?r=US&IR=T",
		"createdUTC":1487844877,"name":"t3_5vpnjl","author":"yeoldgolddigger",
		"title":"Elijah Wood says there's a secret paedophile ring in Hollywood and child actors are 'prey'",
		"score":4329
	};
	const onBackMockFunc = jest.fn();
	onBackMockFunc.mockImplementation(() => {});
	let domTree = null;

	beforeEach(() => {
		domTree = renderer.create(
			<PostView selectedPost={selectedPost} onBack={onBackMockFunc}/>
		);
	});

	test("General rendering", () => {
		const tree = domTree.toJSON();
		expect(tree).toMatchSnapshot();
	});
	test("Checking touchability of \"Back\" button", () => {
		const backButton = domTree.getInstance().backButton;
		expect(typeof backButton.props.onPress).toEqual("function");
		backButton.props.onPress();
		expect(onBackMockFunc).toBeCalled();
	});
	describe("Testing work of alerts", () => {
		beforeEach(() => {
			Alert.alert.mock.calls = [];
		});
		test("canOpenURL == false, error when handling",
			makeTestFunction("Cannot handle this URL", false, true));
		test("canOpenURL == true, no error when handling",
			makeTestFunction(null, true, false));
	});
});
