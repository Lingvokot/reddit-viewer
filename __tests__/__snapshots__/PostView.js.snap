// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PostView General rendering 1`] = `
<View>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#0C42FD",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Back
      </Text>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    Elijah Wood says there's a secret paedophile ring in Hollywood and child actors are 'prey'
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    ID: t3_5vpnjl
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Author: yeoldgolddigger
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Score: 4329
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    External URL:
  </Text>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "blue",
          "textDecorationLine": "underline",
        }
      }
    >
      http://www.businessinsider.com.au/elijah-wood-hollywood-pedophile-ring-2016-5?r=US&IR=T
    </Text>
  </View>
</View>
`;
